### DISCUSSION TITLE
tracking mixed up when instances too close

### DISCUSSION THREAD

**Discussion by ttunja** [talmolab/sleap#935] [Help!]
If we focus on the two mice on the left (with blue marks) once they go over each other the tracking and marking gets mixed up. My question is: In the situation like this, where instances are obviously overlapping how to ensure that  the right tracker ID remains assigned and  the IDs dont get mixed up.

Any suggestions in the tracking configuration to be more strickt about this and take for example relative position in previous frames?

Second question is: Choice of the model and its configuration does not matter right? For this problem its only about configuring tracking correctly?

![C__Users_ttunj_OneDrive_Desktop_labels_whole_mice v001 slp - SLEAP v1 2 6 08_09_2022 11_42_47](https://user-images.githubusercontent.com/60556758/189090829-b648178c-e819-4617-82af-7832443ae43a.png)
![C__Users_ttunj_OneDrive_Desktop_labels_whole_mice v001 slp - SLEAP v1 2 6 08_09_2022 11_43_53](https://user-images.githubusercontent.com/60556758/189090835-d27b7c2c-688c-4161-84c1-f02160171c95.png)

**Comment by talmo** [comment]
Hi @ttunja,

First off: great images! Looks like it'll be a terrific dataset.

Do you mind sharing which tracking parameters you've tried so far to get the results you're currently seeing?

Also, are you just tracking a single node on each animal?

Cheers,

Talmo

**Comment by ttunja** [comment]
Hi @talmo, thanks for the quick feedback.

here is my configuration:

![SLEAP v1 2 6 08_09_2022 20_32_55](https://user-images.githubusercontent.com/60556758/189199240-2403cb72-23a0-48f6-ba7f-5f64bb0e9132.png)


And no, I hope I specified correctly that I am using 4 nodes not one for tracking. (because the 4 nodes for training were head, 2*x´body and tail)

**Comment by talmo** [comment]
That looks pretty good -- you could try using the "flow" method instead of simple, decreasing the frame window to ~5, and using "instance" for similarity.

Do you get a lot of merged skeletons when the animals are too close or do the poses mostly look ok?

**Comment by ttunja** [comment]
Thank you for the suggestions. So I played around a little bit and with the following config most of the conflicts when animals get too close get resolved. (I do have a subquestion: What do you suggest we set "Cull using IoU threshold?" the higher values were better for me)

![SLEAP v1 2 6 09_09_2022 10_21_35](https://user-images.githubusercontent.com/60556758/189306012-12d762a0-7fea-4792-94c1-7f2c8a9e5403.png)

Now with this conflicts get resolved but I get A LOT of empty frame predictions (as you can see in the frames underneath) now as a result of non of the tracks getting predicting for a long time, once it starts predicting back there are mixing ups.
How can one remedy and make sure that every frame MUST have always 4 predictions. Is this enforcable? In this case I suspect it has something to do with the modell too, not only the tracking. I posed a seperate question for this, because I think it will be interesting for the community in the future:  https://github.com/talmolab/sleap/discussions/939


![C__Users_ttunj_OneDrive_Desktop_labels_whole_mice v001 slp - SLEAP v1 2 6 09_09_2022 10_21_10](https://user-images.githubusercontent.com/60556758/189306576-f54be340-dbfc-457b-a012-3f4066f325c4.png)

**Reply by roomrys** [reply]
Hi @ttunja,

If your instances overlap often and by a lot (i.e. positioned on top of each other), then a higher IOU threshold makes sense. The higher the IOU threshold, the less instances will disappear when they overlap.

While there is not way to ensure that SLEAP will predict a minimum number of instances, if you call `sleap-track` through the CLI, you can set the `--peak-threshold` parameter to a lower number (the default is 0.2). Predictions will be removed based on how confident SLEAP is that these predictions are correct. Using a lower `--peak-threshold` value will lower how confident SLEAP needs to be for the predictions to stay. Consequently, you will get more predictions, but they may not be as good as expected. Alternatively, you can train the model with more labeled frames which should also raise the number of predictions returned as confidence increases.

Hope this helps!

Thanks,
Liezl
